extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/CacheScheduler.java

```diff
@@ -0,0 +1,499 @@
+          // via it's own Cleaner as well, when the cache becomes unreachable. Because when somebody forgets to call
+          // entry.close(), it may be harmful to forcibly close the cache, which could still be used, at some
+          // non-deterministic point of time. Cleaners are introduced to mitigate possible errors, not to escalate them.
+          if (calledManually && lastCacheState instanceof VersionedCache) {
+            ((VersionedCache) lastCacheState).cacheHandler.close();
+          }
+        }
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    private void logExecutionError()
+    {
+      if (updaterFuture.isDone()) {
+        try {
+          updaterFuture.get();
+        }
+        catch (ExecutionException ee) {
+          log.error(ee.getCause(), "Error in %s", this);
+        }
+        catch (CancellationException ce) {
+          log.error(ce, "Future for %s has already been cancelled", this);
+        }
+        catch (InterruptedException ie) {
+          Thread.currentThread().interrupt();
+          throw new RuntimeException(ie);
+        }
+      }
+    }
+
+    @Override
+    public String toString()
+    {
+      return asString;
+    }
+  }
+
+  public interface CacheState
+  {}
+
+  public enum NoCache implements CacheState
+  {
+    CACHE_NOT_INITIALIZED,
+    ENTRY_CLOSED
+  }
+
+  public final class VersionedCache implements CacheState, AutoCloseable
+  {
+    final String entryId;
+    final CacheHandler cacheHandler;
+    final String version;
+
+    private VersionedCache(String entryId, CacheHandler cacheHandler, String version)
+    {
+      this.entryId = entryId;
+      this.cacheHandler = cacheHandler;
+      this.version = version;
+    }
+
+    public Map<String, String> getCache()
+    {
+      return cacheHandler.getCache();
+    }
+
+    public String getVersion()
+    {
+      return version;
+    }
+
+    @Override
+    public void close()
+    {
+      log.debug("Closing version [%s] of %s", version, entryId);
+      cacheHandler.close();
+    }
+  }
+
+  private final Map<Class<? extends ExtractionNamespace>, ExtractionNamespaceCacheFactory<?>> namespacePopulatorMap;
+  private final NamespaceExtractionCacheManager cacheManager;
+  private final AtomicLong updatesStarted = new AtomicLong(0);
+  private final AtomicInteger activeEntries = new AtomicInteger();
+
+  @Inject
+  public CacheScheduler(
+      final ServiceEmitter serviceEmitter,
+      final Map<Class<? extends ExtractionNamespace>, ExtractionNamespaceCacheFactory<?>> namespacePopulatorMap,
+      NamespaceExtractionCacheManager cacheManager
+  )
+  {
+    this.namespacePopulatorMap = namespacePopulatorMap;
+    this.cacheManager = cacheManager;
+    cacheManager.scheduledExecutorService().scheduleAtFixedRate(
+        new Runnable()
+        {
+          long priorUpdatesStarted = 0L;
```

Comment:
Thanks, I didn't think about this. `ScheduledExecutorService` doesn't make any guarantees but `ScheduledThreadPoolExecutor` does:

> Successive executions of a task scheduled via scheduleAtFixedRate or scheduleWithFixedDelay do not overlap. While different executions may be performed by different threads, the effects of prior executions happen-before those of subsequent ones.
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledThreadPoolExecutor.html

So I made `NamespaceExtractionCacheManager.scheduledExecutorService()` to return `ScheduledThreadPoolExecutor`.